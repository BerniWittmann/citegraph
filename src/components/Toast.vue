<template>
  <v-snackbar
    class="toast"
    v-model="visible"
    :color="toast.severity"
    :right="true"
    :timeout="toast.delay"
    :top="true"
    @input="close"
  >
    <v-icon dark>
      {{ toast.icon }}
    </v-icon>

    {{ $t(toast.text) }}
    <v-btn
      icon
      @click="close"
    >
      <v-icon>
        mdi-close
      </v-icon>
    </v-btn>
  </v-snackbar>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'
import Toast from '@/models/toast'

@Component
export default class ToastComponent extends Vue {
  @Prop(Toast) readonly toast!: Toast
  visible: boolean = false

  mounted () {
    this.visible = true
  }

  close (): void {
    this.visible = false
    this.$store.dispatch('toasts/removeToast', this.toast)
  }
}
</script>

<style lang="scss">
.toast.v-snack {
  position: relative;
  margin-bottom: 10px;
  margin-right: 10px;
}
</style>
