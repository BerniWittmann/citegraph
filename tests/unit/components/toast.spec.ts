import { shallowMount } from '@vue/test-utils'
import { i18n } from '../setupPlugins'

import ToastComponent from '@/components/Toast.vue'
import Toast, { ErrorToast, InfoToast, SuccessToast, WarningToast } from '@/models/toast'

describe('components/Toast.vue', () => {
  const toasts : Array<Toast> = [
    new SuccessToast('My message'),
    new InfoToast('My message'),
    new WarningToast('My message'),
    new ErrorToast('My message')
  ]
  for (let toast of toasts) {
    describe('the toast is of type ' + toast.severity, () => {
      it('renders', () => {
        const wrapper = shallowMount(ToastComponent, {
          i18n,
          propsData: {
            toast
          }
        })
        expect(wrapper.html()).toMatchSnapshot()
      })

      it('renders the message', () => {
        const wrapper = shallowMount(ToastComponent, {
          i18n,
          propsData: {
            toast
          }
        })
        expect(wrapper.text()).toContain('My message')
      })

      it('shows the toast on mount', () => {
        const wrapper = shallowMount(ToastComponent, {
          i18n,
          propsData: {
            toast
          }
        })
        // @ts-ignore
        expect(wrapper.vm.visible).toBeTruthy()
      })

      it('can close the toast', () => {
        const dispatch = jest.fn()
        const wrapper = shallowMount(ToastComponent, {
          i18n,
          propsData: {
            toast
          },
          mocks: {
            $store: {
              dispatch
            }
          }
        })
        const btn = wrapper.find('v-btn-stub')
        expect(btn.exists()).toBeTruthy()

        // @ts-ignore
        wrapper.vm.close()

        // @ts-ignore
        expect(wrapper.vm.visible).toBeFalsy()
        expect(dispatch).toHaveBeenCalledWith('toasts/removeToast', toast)
      })
    })
  }
})
