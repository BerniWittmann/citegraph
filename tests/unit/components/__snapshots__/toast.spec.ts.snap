// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Toast.vue the toast is of type error renders 1`] = `
<v-snackbar-stub color="error" top="true" right="true" timeout="6000" class="toast" value="true">
  <v-icon-stub dark="">
    mdi-alert-circle-outline
  </v-icon-stub>

  My message
  <v-btn-stub tag="button" activeclass="" icon="true" type="button">
    <v-icon-stub>
      mdi-close
    </v-icon-stub>
  </v-btn-stub>
</v-snackbar-stub>
`;

exports[`components/Toast.vue the toast is of type info renders 1`] = `
<v-snackbar-stub color="info" top="true" right="true" timeout="4000" class="toast" value="true">
  <v-icon-stub dark="">
    mdi-information-outline
  </v-icon-stub>

  My message
  <v-btn-stub tag="button" activeclass="" icon="true" type="button">
    <v-icon-stub>
      mdi-close
    </v-icon-stub>
  </v-btn-stub>
</v-snackbar-stub>
`;

exports[`components/Toast.vue the toast is of type success renders 1`] = `
<v-snackbar-stub color="success" top="true" right="true" timeout="3000" class="toast">
  <v-icon-stub dark="">
    mdi-check-circle-outline
  </v-icon-stub>

  My message
  <v-btn-stub tag="button" activeclass="" icon="true" type="button">
    <v-icon-stub>
      mdi-close
    </v-icon-stub>
  </v-btn-stub>
</v-snackbar-stub>
`;

exports[`components/Toast.vue the toast is of type warning renders 1`] = `
<v-snackbar-stub color="warning" top="true" right="true" timeout="5000" class="toast" value="true">
  <v-icon-stub dark="">
    mdi-alert
  </v-icon-stub>

  My message
  <v-btn-stub tag="button" activeclass="" icon="true" type="button">
    <v-icon-stub>
      mdi-close
    </v-icon-stub>
  </v-btn-stub>
</v-snackbar-stub>
`;
